pkg/query-service/app/logparsingpipeline/postablePipeline.go

```diff
@@ -0,0 +1,178 @@
+}
+
+// PostablePipeline captures user inputs in setting the pipeline
+
+type PostablePipeline struct {
+	Id          string                   `json:"id"`
+	OrderId     int                      `json:"orderId"`
+	Name        string                   `json:"name"`
+	Alias       string                   `json:"alias"`
+	Description string                   `json:"description"`
+	Enabled     bool                     `json:"enabled"`
+	Filter      string                   `json:"filter"`
+	Config      []model.PipelineOperator `json:"config"`
+}
+
+// IsValid checks if postable pipeline has all the required params
+func (p *PostablePipeline) IsValid() *model.ApiError {
+	if p.OrderId == 0 {
+		return model.BadRequestStr("orderId with value > 1 is required")
+	}
+	if p.Name == "" {
+		return model.BadRequestStr("pipeline name is required")
+	}
+
+	if p.Alias == "" {
+		return model.BadRequestStr("pipeline alias is required")
+	}
+
+	if p.Filter == "" {
+		return model.BadRequestStr("pipeline filter is required")
+	}
+
+	// check the expression
+	_, err := expr.Compile(p.Filter, expr.AsBool(), expr.AllowUndefinedVariables())
+	if err != nil {
+		return model.BadRequestStr(fmt.Sprintf("filter for pipeline %v is not correct: %v", p.Name, err.Error()))
+	}
+
+	idUnique := map[string]struct{}{}
+	outputUnique := map[string]struct{}{}
+
+	l := len(p.Config)
+	for i, op := range p.Config {
+		if op.OrderId == 0 {
+			return model.BadRequestStr("orderId with value > 1 is required in operator")
+		}
+		if op.ID == "" {
+			return model.BadRequestStr("id of an operator cannot be empty")
+		}
+		if op.Type == "" {
+			return model.BadRequestStr("type of an operator cannot be empty")
+		}
+		if i != (l-1) && op.Output == "" {
+			return model.BadRequestStr(fmt.Sprintf("Output of operator %s cannot be nil", op.ID))
+		}
+		if i == (l-1) && op.Output != "" {
+			return model.BadRequestStr(fmt.Sprintf("Output of operator %s should be empty", op.ID))
+		}
+
+		if _, ok := idUnique[op.ID]; ok {
+			return model.BadRequestStr("duplicate id cannot be present")
+		}
+		if _, ok := outputUnique[op.Output]; ok {
+			return model.BadRequestStr("duplicate output cannot be present")
+		}
+
+		if op.ID == op.Output {
+			return model.BadRequestStr("id and output cannot be same")
+		}
+
+		err := isValidOperator(op)
+		if err != nil {
+			return err
+		}
+
+		idUnique[op.ID] = struct{}{}
+		outputUnique[op.Output] = struct{}{}
+	}
+	return nil
+}
+
+func isValidOperator(op model.PipelineOperator) *model.ApiError {
+	valueErrStr := "value should have prefix of body, attributes, resource"
+
+	switch op.Type {
+	case "json_parser":
+		if op.ParseFrom == "" && op.ParseTo == "" {
+			return model.BadRequestStr(fmt.Sprintf("parse from and parse to of %s json operator cannot be empty", op.ID))
+		}
+	case "grok_parser":
+		if op.Pattern == "" {
+			return model.BadRequestStr(fmt.Sprintf("pattern of %s grok operator cannot be empty", op.ID))
```

Comment:
For every single one of these `model.BadRequestStr`, I would create a [errors.New](https://golang.org/pkg/errors/#New) and use.